src/dbnode/persist/fs/retriever.go

```diff
@@ -372,11 +371,32 @@ func (r *blockRetriever) fetchBatch(
 ) {
 	var seeker ConcurrentDataFileSetSeeker
 	defer func() {
+		filteredReqs := allReqs[:0]
 		// Make sure requests are always fulfilled so if there's a code bug
 		// then errSeekNotCompleted is returned because req.success is not set
 		// rather than we have dangling goroutines stacking up.
 		for _, req := range allReqs {
-			req.onDone()
+			if req.waitOnFetch == nil {
+				req.onDone()
+				continue
+			}
+
+			filteredReqs = append(filteredReqs, req)
+		}
+
+		allReqs = filteredReqs
+		for len(allReqs) > 0 {
+			filteredReqs = filteredReqs[:0]
+			for _, req := range allReqs {
+				select {
+				case <-req.waitOnFetch:
+					req.onDone()
+				default:
+					filteredReqs = append(filteredReqs, req)
+				}
+			}
+
+			allReqs = filteredReqs
```

Comment:
Won't this result in a busy wait? I.e. spinning on 100% CPU until `allReqs` becomes empty.
At the minimum, a short sleep would help, but maybe this can be rewritten for example in terms of a `WaitGroup`?

Reply:
Yeah I guess I was overoptimizing for finishing requests as soon as possible, removing the select from here to block on the channel instead.